File: simple/javavulnweb/src/main/java/com/vulnweb/common/VulnParent.java
package com.vulnweb.common;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;

/**
 * 公共父类，集中存放所有漏洞核心方法
 * 包含命令执行、代码执行、SSRF三类高危漏洞
 */
public class VulnParent {
    /**
     * 命令执行漏洞点
     * 使用 Runtime.getRuntime().exec() 执行外部命令
     * @param cmd 要执行的命令
     * @return 命令执行结果
     */
    public String cmdExec(String cmd) {
        try {
            // 漏洞点：直接执行用户传入的命令，未做任何过滤
            Process process = Runtime.getRuntime().exec(cmd);
            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            StringBuilder output = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }
            reader.close();
            return output.toString();
        } catch (Exception e) {
            return "命令执行错误: " + e.getMessage();
        }
    }

    /**
     * 代码执行漏洞点
     * 使用 ScriptEngine 执行JavaScript代码
     * @param script 要执行的脚本
     * @return 脚本执行结果
     */
    public String codeExec(String script) {
        try {
            // 漏洞点：直接执行用户传入的脚本，未做任何过滤
            ScriptEngineManager manager = new ScriptEngineManager();
            ScriptEngine engine = manager.getEngineByName("JavaScript");
            Object result = engine.eval(script);
            return result.toString();
        } catch (Exception e) {
            return "代码执行错误: " + e.getMessage();
        }
    }

    /**
     * SSRF漏洞点
     * 使用 HttpURLConnection 发起HTTP请求
     * @param url 目标URL
     * @return HTTP响应内容
     */
    public String ssrf(String url) {
        try {
            // 漏洞点：直接请求用户传入的URL，未做任何过滤
            URL targetUrl = new URL(url);
            HttpURLConnection connection = (HttpURLConnection) targetUrl.openConnection();
            connection.setRequestMethod("GET");
            BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line).append("\n");
            }
            reader.close();
            return response.toString();
        } catch (Exception e) {
            return "SSRF错误: " + e.getMessage();
        }
    }
}

File: simple/scalavulnweb/app/models/common/VulnParent.scala
package models.common

// 父类：包含命令执行、代码执行、SSRF漏洞方法
// 所有漏洞点通过子类继承、特质混入形成多层调用链
class VulnParent {

  // 【漏洞点】命令执行 - 硬编码场景（无外部输入）
  def cmdExecHardcoded(): String = {
    val cmd = "echo vuln_hardcoded_cmd"
    val runtime = Runtime.getRuntime
    // 【漏洞】直接执行硬编码命令，无任何过滤
    val process = runtime.exec(cmd)
    val result = scala.io.Source.fromInputStream(process.getInputStream).mkString
    process.waitFor()
    result
  }

  // 【漏洞点】命令执行 - 外部可控参数（通过子类/特质混入传递）
  def cmdExecWithParam(input: String): String = {
    val runtime = Runtime.getRuntime
    // 【漏洞】直接拼接外部参数执行命令，命令注入风险
    val process = runtime.exec(Array("sh", "-c", input))
    val result = scala.io.Source.fromInputStream(process.getInputStream).mkString
    process.waitFor()
    result
  }

  // 【漏洞点】代码执行 - 硬编码场景
  def codeExecHardcoded(): String = {
    import javax.script.{ScriptEngine, ScriptEngineManager}
    val manager = new ScriptEngineManager()
    val engine = manager.getEngineByName("js")
    // 【漏洞】直接执行硬编码脚本代码
    val code = "1 + 2 * 3"
    val result = engine.eval(code).toString
    result
  }

  // 【漏洞点】代码执行 - 外部可控参数
  def codeExecWithParam(input: String): String = {
    import javax.script.{ScriptEngine, ScriptEngineManager}
    val manager = new ScriptEngineManager()
    val engine = manager.getEngineByName("js")
    // 【漏洞】直接执行外部传入的脚本代码，代码注入风险
    val result = engine.eval(input).toString
    result
  }

  // 【漏洞点】SSRF - 硬编码场景
  def ssrfHardcoded(): String = {
    // 【漏洞】直接请求硬编码的内网地址
    val url = "http://127.0.0.1:8080/internal"
    val connection = new java.net.URL(url).openConnection()
    connection.setConnectTimeout(3000)
    val stream = connection.getInputStream
    val result = scala.io.Source.fromInputStream(stream).mkString
    stream.close()
    result
  }

  // 【漏洞点】SSRF - 外部可控参数
  def ssrfWithParam(input: String): String = {
    // 【漏洞】直接请求外部传入的URL，SSRF风险
    val connection = new java.net.URL(input).openConnection()
    connection.setConnectTimeout(3000)
    val stream = connection.getInputStream
    val result = scala.io.Source.fromInputStream(stream).mkString
    stream.close()
    result
  }
}
